src/server/Demo.UnitTests/TestControllerTests.cs

```diff
@@ -18,11 +19,11 @@ public TestControllerTests()
         }
 
         [Fact]
-        public async void TestController_GetAction_MustReturnOkObjectResult()
+        public void TestController_GetAction_MustReturnOkObjectResult()
```

Comment:
I am rather confused by this method.
First, the type of `result` is well-known - defined by the signature of `_controller.Test()`. The check at line 26 seems redundant.
Second, `Assert.IsType<OkObjectResult>` only works because `OkObjectResult` is a sealed class (implementation detail IMHO). I would still go with `Assert.IsAssignableFrom<OkObjectResult>` instead.
Last but not least, the name of the method is confusing to me. Looking at `_controller.Test()` implementation, there's clearly a case when the method doesn't return `OkObjectResult`. I believe the "test case" here is to verify that _for an empty `Foo` repository_ the method returns OK - would like to see that fact projected into the test name. I would also tend to check that it returns an empty collection of `Foo`s.

Reply:
The first test at line 26 is an artifact I forgot in there when the controller was returning IActionResult. I will remove it.
Regarding IsType vs IsAssignableFrom, I was taught to keep the tests as tight as possible where extensibility is unlikely (which with ObjectResult derivates it is).
Re: Naming. Yes there is a case where the returned type will not be an OkObjectResult. But that is not the goal of that single test - the test aims at a scenario where the returned type must be OkObjectResult. Basically, the test method does not say that the controller action must always return an OK result, but that in this scenario, it must. This is a matter of naming convention we want to set up. Perhaps I we could exclude the imperative (Must) and rename it to ReturnsOkObjectResult?

Reply:
IsType vs IsAssignableFrom - for me it looks more like your test depends on an implementation detail, which makes it fragile in case the real code is refactored. Because, what we actually wanted to verify was the HTTP Status code 200 and not the fact that the implementation achieves it by means of using the sealed class `OkObjectResult`. But I see your point about stuff unlikely to change. So, OK, screw it, keep it like it is now.

Now, I would argue that the method name actually currently does say that the controller action must always return an OK and omitting _Must_ will not help. And I can think of different yet valid test scenarios where OkObjectResult would be returned as well (e.g. having 5 `Foo` items in the Repository). So, the problem here is exactly the fact that the name of the method only tells you what the result should look like, but doesn't actually say _which_ scenario it is testing (we know it is testing the _empty Repository_ scenario, but we only know it by looking at its implementation).
Plus, the prefix `TestController_` is actually redundant (the name of the class tested is apparent from `TestControllerTests`).
So, instead of omitting _Must_ I would rather come up with a name like `EmptyRepository_GetAction_ReturnsOK` or something.
In general, the test naming convention could be: \<precondition\>`_`\<action\>`_`\<expected result\>.
What do you think about it?